### PR TITLE
Late ack record container

### PR DESCRIPTION
**Pull request by lbudai**
see: https://github.com/balabit/syslog-ng/pull/2785#issuecomment-503014960

This is the PR I mentioned: `LateAckTracker` does not contain any fixed-size related code anymore.
(I have a working branch where I already implemented the list-based `LateAckTrackerContainerDynamic` class, so I think the interface is stable.)

* note: I also moved all the ack-tracker related code (except Ringbuffer, which is a general 'container') to a separate directory.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/ack-tracker/late_ack_record_container_static.c [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  self->pending = NULL;
+}
+
+static void
+_drop(LateAckRecordContainer *s, gsize n)
+{
+  StaticLateAckRecordContainer *self = (StaticLateAckRecordContainer *)s;
+  int i;
+  LateAckRecord *ack_rec;
+  n = MIN(ring_buffer_count(&self->ack_records), n);
```
Comment (MrAnno) [inline lib/ack-tracker/late_ack_record_container_static.c]: In my understanding it is a bug if `n` is greater than `ring_buffer_count`.
Shouldn't we change this to an assert (`g_assert(n <= ring_buffer_count(&self->ack_records))`)?

File: lib/ack-tracker/late_ack_record_container.h [hunk 2]
```diff
@@ -53,6 +55,8 @@ late_ack_record_destroy(LateAckRecord *self)
     self->bookmark.destroy(&(self->bookmark));
 }
 
+LateAckRecordContainer *late_ack_record_container_static_new(gsize size);
```
[the remark this answers is not in the document]
Comment (lbudai) [reply lib/ack-tracker/late_ack_record_container.h]: I would not create a separate header (it's not a bad thing when you have one header and multiple implementation files).

File: lib/ack-tracker/late_ack_record_container_static.c (hunk 1 above)
Comment (lbudai) [reply lib/ack-tracker/late_ack_record_container_static.c]: ok

File: lib/ack-tracker/late_ack_record_container_static.c [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+static void
+_free(LateAckRecordContainer *s)
+{
+  StaticLateAckRecordContainer *self = (StaticLateAckRecordContainer *)s;
+
+  guint32 count = ring_buffer_count(&self->ack_records);
+  _drop(s, count);
+  ring_buffer_free(&self->ack_records);
```
[the remark this answers is not in the document]
Comment (lbudai) [reply lib/ack-tracker/late_ack_record_container_static.c]: thx!

File: lib/ack-tracker/late_ack_record_container.h [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+typedef struct _LateAckRecordContainer LateAckRecordContainer;
+
+struct _LateAckRecordContainer
+{
+  gboolean (*is_empty)(const LateAckRecordContainer *s);
+  LateAckRecord *(*request_pending)(LateAckRecordContainer *s);
+  void (*store_pending)(LateAckRecordContainer *s);
+  void (*drop)(LateAckRecordContainer *s, gsize n);
+  LateAckRecord *(*at)(const LateAckRecordContainer *s, gsize idx);
+  void (*free)(LateAckRecordContainer *s);
```
[the remark this answers is not in the document]
Comment (lbudai) [reply lib/ack-tracker/late_ack_record_container.h]: well... I never liked `free_fn`... but will change

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by lbudai** [comment]
@kira-syslogng  : retest this please

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by lbudai** [comment]
@kira-syslogng : retest this please

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by bazsi** [comment]
On some platforms free() is a macro, which will cause this to break.

So I would have preferred free() myself.

I am not sure that macro breakage still holds though.

Budai Laszlo <notifications@github.com> ezt írta (időpont: 2019. jún. 19.,
Sze 14:46):

**Comment by lbudai** [comment]
@bazsi: I'm ok with using free_fn, just always forget it and first versions of my d'tors are free :)

**Pull request merged**